x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java

```diff
@@ -63,24 +64,26 @@ public Weight doCache(Weight weight, QueryCachingPolicy policy) {
      */
     static boolean cachingIsSafe(Weight weight, IndicesAccessControl.IndexAccessControl permissions) {
         // support caching for common queries, by inspecting the field
-        // TODO: If in the future there is a Query#extractFields() then we can do a better job
-        Set<String> fields = new HashSet<>();
         try {
-            FieldExtractor.extractFields(weight.getQuery(), fields);
-        } catch (UnsupportedOperationException ok) {
-            // we don't know how to safely extract the fields of this query, don't cache.
-            return false;
-        }
+            weight.getQuery().visit(new QueryVisitor() {
+                @Override
+                public QueryVisitor getSubVisitor(BooleanClause.Occur occur, Query parent) {
+                    return this; // we want to use the same visitor for must_not clauses too
+                }
 
-        // we successfully extracted the set of fields: check each one
-        for (String field : fields) {
-            // don't cache any internal fields (e.g. _field_names), these are complicated.
-            if (field.startsWith("_") || permissions.getFieldPermissions().grantsAccessTo(field) == false) {
-                return false;
-            }
+                @Override
+                public boolean acceptField(String field) {
+                    // don't cache any internal fields (e.g. _field_names), these are complicated.
+                    if (field.startsWith("_") || permissions.getFieldPermissions().grantsAccessTo(field) == false) {
+                        throw new UnsupportedOperationException();
+                    }
+                    return super.acceptField(field);
+                }
+            });
+        } catch (UnsupportedOperationException e) {
+            return false;
```

Comment:
I've verified that for the query types we previously checked for cacheability in the FLS context, the new strategy using the visitor is equivalent.

But as has been mentioned in the PR already, I'm reluctant to cast aside the allow list of cacheable queries. I would prefer that, in addition to the field verifier using the visitor, also keep the allow list of query types that are cacheable. For example, I'm not sure if all the query types that only call `QueryVisitor#visitLead` and don't call `QueryVisitor#acceptField` are cacheable (they might be, I just don't know enough to be sure of it).

Reply:
I think I am not seeing how combining the two approaches would work, that is why I am thinking that if you want an allow list, query visitors won't help. I tried this in previous iterations and it did not work, but I may be missing something. Would you mind taking this over? You can freely push new commits to my branch.

Reply:
The tricky part is detecting when a query calls `visitLeaf` without having called `acceptField` first.  I think MoreLikeThisQuery should actually check its allowed fields first, for example.  We can have a look at all our query implementations and check that they are calling `acceptField` correctly.

Reply:
> Would you mind taking this over? You can freely push new commits to my branch.

Okay, will try it tomorrow.

Reply:
no rush, I don't mean to indicate that this is urgent, just that I'd prefer to hand this over to you and your team if there is interest in getting it merged. Thanks!

Reply:
Looking at this again... FieldExtractor is called from within a query cache implementation, and by the time we get to it queries will have been through a rewrite phase already.  So queries like MoreLikeThis will have been rewritten into a BooleanQuery containing a bunch of term queries, which will be handled correctly here.

From a quick look through the Query implementations we have, they either a) correctly implement `visit`, b) rewrite to a core Query that implements `visit` or c) mark themselves as not cacheable.  So I think I'm happy with this implementation.

Reply:
I've pushed the allow list for query types: [87f0eb8](https://github.com/elastic/elasticsearch/pull/93390/commits/87f0eb8bfd60481ba84866f9c647dbbc866a8aaf) .

Overall the implementation still trusts that the visitor API is implemented correctly by queries, i.e. that queries call the visitor methods (which I think is equivalent to trusting that the old `FieldExtractor` works in the same way), but also guards against implementations that omit to call the `acceptField` method.
This also cover for the unlikely case that some queries should not be cached even if they don't refer to doc fields that the user is not allowed to access. But maybe this is never the case?

Reply:
Sorry for being pedantic, and thanks for taking this over. I don't understand what the allowlist cover for. Is it because we only want to cache known queries, or because the visit implementation may be wrong, or both?

My thinking is that we were doing instanceof checks before because there was no generic visitor API to iterate over the query tree and retrieve fields that queries were searching against, not necessarily to have more control over what query types get cached, but I may be wrong.

Btw I found this old issue: is it still valid / related to this work?

Reply:
> I don't understand what the allowlist cover for. Is it because we only want to cache known queries, or because the visit implementation may be wrong, or both?

**The requirement is that queries that access disallowed fields (for some user) should not be cached (for all users).**
So I think that if we had an API to get the fields for all query types, we wouldn't need an allowlist of query types.

So the question is if the visitor API can play that role.
To me it looks decoupled from the query evaluation logic (meaning that there's a higher chance to be implemented differently from how the fields are computed when evaluating a query).
Hence my suggestion that we play it safe and use the visitor API for an allowlist of query types that we know it works correctly (today) to extract the field names. If it weren't for a security use-case, where the consequences of a bug would be less serious, I would for sure not suggest an allowlist.

This is just my personal suggestion, and I'm probably downplaying the robustness of the visitor API, and how frequent new query types get added.

So if the search folks says the visitor API is good to be used to extract the field names for all the query types, then that fits the requirement. The original hardcoded implementation is not foolproof either, as technically it's possible the field name extraction logic breaks for existing query types after a refactoring.

Reply:
Thanks for your patience Albert, I think the allowlist adds some robustness, but using the visitor you still need the query to interact with the visitor in some predictable way, which you can't guarantee. Also, we will only deal with rewritten queries here, hence quite low-level and custom queries matter less. I am tempted to say: the visitor is the generic mechanism that we are supposed to get the field names from and we should be able to rely on it. If a query does not follow the visitor contract, it is a buggy query that needs fixing.

Reply:
If the visitor API can be used as an API to retrieve all the fields for all the query types, I don't see any concerns Securiy-wise to droping the allowlist.
For FLS, in this context, we only need to verify and not cache queries that refer to fields that the user has FLS permissions on.

Reply:
That is the case Albert, does it mean that you are ok with removing the allowlist then? I think we can keep things simple and rely on the visit implementation.

Reply:
Okay @javanna , lets drop the `acheableQueryTypes` Set and rely exclusively on the query visitor to extract the field names.

Reply:
thanks, I pushed an update